### ISSUE TITLE
Haddock gets confused with a -- * comment

### ISSUE THREAD

**Issue by LeventErkok** [haskell/haddock#798]
I suspect this might be a GHC issue as well, but Haddock is choking on this following:

```haskell
bug = 1
    -- * fails
```

I get:

```
$ haddock --html a.hs
Haddock coverage:

a.hs:2:5: error: parse error on input ‘-- * fails’
```

Inserting some extra space between `--` and `*` solves the problem. But surely this is something haddock/ghc should be able to handle?

**Comment by alexbiehl** [comment]
Yeah, thats ghc again. Could you file upstream? We shouldnt fail for invalid comments.

**Comment by LeventErkok** [comment]
Sure, filed here: https://ghc.haskell.org/trac/ghc/ticket/15029

**Comment by GeorgeCo** [comment]
Hi, Can you explain why this is a ghc issue? I couldn't find anything in the User's Guide and in the Haskell 2010 report all I found is: 

2.3 Comments
Comments are valid whitespace.
An ordinary comment begins with a sequence of two or more consecutive dashes (e.g. --) and extends to the following newline. The sequence of dashes must not form part of a legal lexeme. For example, “-->” or “|--” do not begin a comment, because both of these are legal lexemes; however “--foo” does start a comment.
A nested comment begins with “{-” and ends with “-}”. No legal lexeme starts with “{-”; hence, for example, “{---” starts a nested comment despite the trailing dashes.
The comment itself is not lexically analysed. Instead, the first unmatched occurrence of the string “-}” terminates the nested comment. Nested comments may be nested to any depth: any occurrence of the string “{-” within the nested comment starts a new nested comment, terminated by “-}”. Within a nested comment, each “{-” is matched by a corresponding occurrence of “-}”.
In an ordinary comment, the character sequences “{-” and “-}” have no special significance, and, in a nested comment, a sequence of dashes has no special significance.
Nested comments are also used for compiler pragmas, as explained in Chapter 12.
If some code is commented out using a nested comment, then any occurrence of {- or -} within a string or within an end-of-line comment in that code will interfere with the nested comments.

**Comment by harpocrates** [comment]
Haddock doesn't actually parse the source files directly, it goes through GHC (specifically, it depends on `ghc`-the-library). This means that GHC must differentiates `-- * ...` from `-- ...` (at the lexer-level, see https://github.com/ghc/ghc/blob/a5373c1fe172dee31e07bcb7c7f6caff1035e6ba/compiler/parser/Lexer.x#L770-L777).

Ideally, GHC would not throw parse errors for this sort of thing, hence this bug report.
